sjasm/io_snapshots.cpp

```diff
@@ -145,4 +145,89 @@ int SaveSNA_ZX(char* fname, word start) {
 	return 1;
 }
 
+int SaveSNA_CPC(char* fname, word start) {
+	// for Lua
+	if (!DeviceID) {
+		Error("[SAVESNA] Only for real device emulation mode.");
+		return 0;
+	}
+	else if (!IsAmstradCPCDevice(DeviceID)) {
+		Error("[SAVESNA] Device must be AMSTRADCPC464.");
+		return 0;
+	}
+
+	FILE* ff;
+	if (!FOPEN_ISOK(ff, fname, "wb")) {
+		Error("[SAVESNA] Error opening file", fname, FATAL);
+		return 0;
+	}
+
+	// format:  http://cpctech.cpc-live.com/docs/snapshot.html
+	constexpr int SNA_HEADER_SIZE = 256;
+	const char magic[8] = { 'M', 'V', ' ', '-', ' ', 'S', 'N', 'A' };
+	// basic rom initialized pens
+	const byte ga_pens[17] = { 0x04, 0x0A, 0x13, 0x0C, 0x0B, 0x14, 0x15, 0x0D, 0x06, 0x1E, 0x1F, 0x07, 0x12, 0x19, 0x04, 0x17, 0x04 };
+	// crtc set to standard screen @ $C000
+	const byte crtc_defaults[18] = { 
+		0x3F,		// h total
+		0x28,		// h displayed
+		0x2E,		// h sync pos
+		0x8E,		// h/v sync widths
+		0x26,		// v total (height)
+		0x00,		// v adjust
+		0x19,		// v displayed (height)
+		0x1E,		// v sync pos
+		0x00,		// interlace & skew
+		0x07,		// max raster
+		0x00, 0x00, // cursor start
+		0x30, 0x00, // display (xxPPSSOO) 
```

Comment:
One thing that is worth discussing / deciding here is if we want to move SAVESNA into a separate directive, such as OPENCPCSNA and allow customization of some of these defaults; notably the PEN settings and display address may be something a user wants to configure.

I figured implement the simplest thing and change if we think it's needed

Reply:
from the used DEVICE it is clear which `SAVESNA` implementation has to be used, so in this regard I guess the end user doesn't really care about the name collision and the CPC `SAVESNA` may be custom-tailored as much as you want.

But thinking about the "searching documentation" use case, maybe it may be worth to give it custom mnemonics too, so users will not read ZX SAVESNA docs by accidents... about the open/close way like NEX file has vs single SAVESNA (or CPCSAVE or something like that)... the single directive is often easier to use, often outweighing the details, for example ZX SNA can't be configured to have interrupts enabled or disable (although it's trivial to post-patch the outputted file directly form asm source, as it's not encrypted or compressed). 99% of ZX snapshots are fine with defaults in those extra fields. If you know CPC developers like to customise snapshots more, it may be worth to do some open->close stuff like NEX, including full config.
(but generally on ZX the snapshot format is kinda frown upon, it's nice for quick-hack development, but final release should be rather proper TAP image, just dumping whole machine into snap is considered a bit lame... no idea how CPC scene feels about that... if it's similar, then I guess fine configuring of quick-hack format is not very valuable).

Reply:
Yeah, I agree. 

The snapshot format has similar principles; except it's safer on the CPC emulators (no stack modification). Most people distribute stuff as TZX / CDT format (CDT is just TZX by another name). SNA here is used in the same way; eg: hack a build together for dev/testing. I agree though, I'll move the directive to SAVECPC or similar to avoid confusion - will have to have a think about if we care about having all the tweakable values or not. I know I don't need them as I set them on program init :)

I'm also working on adding some simple TZX support for both Speccy and CPC and that follows the SAVENEX style directives.